File: current_daemon.rb
require "serialport"
require "iconv"

def init
	port_str = "/dev/ttyUSB0"  
	baud_rate = 9600
	data_bits = 8
	stop_bits = 1
	parity = SerialPort::NONE
	@rising_edge=0;
	@start_brew=0
	@finish_brew=0
	if ARGV.length > 0
		@sp = File.open(ARGV[0],"r")
		@test=true
	else
		@sp = SerialPort.new(port_str, baud_rate, data_bits, stop_bits, parity)
		@test=false
	end 
end

def tweet(what)
	if @test
		puts "#{what}"
	else
		`twurl -d "status=#{what}" /1/statuses/update.xml`
	end
end

def sample
	if @test
		line=@sp.gets.split
		return [line[0].to_f,line[1].to_f] 
	end

	while true 
		line=@sp.gets
		unless line.nil?
			ic = Iconv.new('UTF-8//IGNORE', 'UTF-8')
			currents = ic.iconv(line)
			a,b =  currents.split
			if a == b 
				return [Time.now.to_f,a.to_f]
			end
		end
	end
end

#TODO detect sample rate?
#assume 1 second for now
def main
	rising_edge=0
	heating=0
	while true
	current=sample
		if current[1] > 0.5 
			if heating == 0 
				puts "#{current[0]} #{heating}"
			end
			heating+=1
			rising_edge = true
			falling_edge = false
		end
		if current[1] < 0.25
			if heating >=1
				falling_edge=true
				rising_edge=false
				puts "#{current[0]} #{heating}"
				if heating > 500
					tweet("#{Time.now} Fresh Pot")
				end
				heating=0
			end
		end
	end
end



#loop till exception?
init
main

@sp.close                       

File: current_daemon.c
#include <sys/types.h>
#include <stdlib.h>
#include <iostream>
#include <time.h>
#include <stdio.h> 
#include <string.h>
#include <unistd.h>
#include <fcntl.h> 
#include <errno.h> 
#include <termios.h>
#include <time.h>   

using namespace std;
#define FRESH_THRESHHOLD 1500
//#define TEST 1

char *buffer,*bufptr;
FILE *input;

#ifdef __linux__
//arduino millis() like
unsigned long millis(){
	struct timespec ts;
	clock_gettime(CLOCK_REALTIME, &ts);
	return((unsigned long)(ts.tv_sec * 1000.0 + (ts.tv_nsec/1000000.0)));
}	
#endif

void notify() {
	int rc=system("/bin/bash ./notify.sh");
	if(rc !=0){
		std::cout << "system returned " << rc << endl;
	}
}

void tweet(const char *what) {
	char tweet_str[141+50];
	time_t rawtime;
	struct tm * timeinfo;

	time ( &rawtime );
	timeinfo = localtime ( &rawtime );
	char *timestamp=asctime (timeinfo) ;
	timestamp[strlen(timestamp)-1]='\0';

	sprintf(tweet_str,
			"twurl -d \"status=%s %s\" /1/statuses/update.xml > /dev/null",
			timestamp, what);
	int rc=system(tweet_str);
	if(rc !=0){
		std::cout << "system returned " << rc;
	}
}

FILE *open_serial_port(){
#ifdef TEST 
	input = fopen("./testsamples.txt", "r");
#else
	int fd;
	FILE *input;
	struct termios termattr;
	speed_t baudRate;
	input = fopen("/dev/ttyUSB0", "r");
	fd = fileno(input);
	tcgetattr(fd, &termattr);
	baudRate = cfgetispeed(&termattr);
	if (baudRate != B9600) {
		cfsetispeed(&termattr, B9600);
		tcsetattr(fd, TCSANOW, &termattr);
	}
#endif
	return input;
}

int sample(FILE *input) {
	char buff[21]; 
	int a=0,b=0;
	char *pb;
#ifdef TEST
	usleep(1000); 
#endif
	while(1){
		char* rc=fgets(buff,20,input);
		a=atoi(buff);
		pb = strchr(buff,' ');
		if(pb != NULL) {
			b=atoi(pb);
		}
		if(a == b) {
			return a;
		}
		usleep(1000); //prevent 100% cpu usage on gets fail
	}
}


int main() {
	int current;
	int	heating=0;
	unsigned long start_heat;
	struct timespec ts;

	input = open_serial_port();

	while(1) {
		current = sample(input);
		if(current >= 1){  
			if(heating == 0 ){
				start_heat = millis();
				std::cout << "Rising " << start_heat << std::endl;
				fflush(stdout);
			}
			else
			{
				std::cout << "Heating "  << millis() << " " << (millis()-start_heat) << std::endl ;
			}
			heating++;
		}
		if(current < 1){
			if(heating >= 1){
				std::cout << "Falling " << millis() << " " << (millis()-start_heat) << std::endl  ;
				fflush(stdout);
				if(heating > FRESH_THRESHHOLD){
					notify();
				}
				heating=0;
			}
		}

	}
	fclose(input);
	return 0;
}



